### PR TITLE
Rework namespace-limits handing

### PR DESCRIPTION
**Pull request by ahadas**
**What this PR does / why we need it**:
Change the current handling of namespace's LimitRanges:
1. We do not look for the LimitRange in the namespace with the minimal value of memory-limit as it is not what Kubernetes does + there is no good reason to prioritize that with the minimal value of memory-limit over the one with the minimal value of cpu-limit, for instance.
2. We apply not only the memory-default but also cpu-default and both memory- and cpu- defaultRequest.

**Special notes for your reviewer**:
Solving the potential issue discussed in #2137 about memory defaultRequest will be done separately.
This PR copies some code from Kubernetes (see references within the code) - it would be great to reuse the code in Kubernetes but it's a bit complex at the moment as it resides within an admission plugin within the main Kubernetes repo.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by rmohr** [comment]
@ahadas thanks! Will have a closer look as soon as I can. It looks very good. Can you adjust our functional test, since we now set more values?

**Comment by ahadas** [comment]
@rmohr , sure, but I'd like to check the memory issue we discussed in #2137 first: it seems that the default request for memory that is set by `pkg/api/v1/defaults.go` is done after the VMI is mutated - so that's good for a case in which the VMI is sent to the apiserver. 
But I'm surprised by the result of passing a VM to the apiserver and then starting it using virtctl - the namespace-limits are not being applied. Could it be that the VMI does not go through the apiserver in that case and therefore the mutating webhook is not applied?

**Comment by ahadas** [comment]
@rmohr adjusted the functional tests and added missing validations for cpu request & limit in the validating-webhook. let's think about the memory related issues separately.

**Comment by rmohr** [comment]
@ahadas I checked, the mutating webhook sets the defaults already. that should work as expected. I just noticed that some defaults are now also set directly inside the mutationg webhook, and not all are specified in `defaults.go`. Some cleanup necessary, but that should not bother this PR in general.

**Comment by ahadas** [comment]
@rmohr I traced the cause for the problem I had - the VM was defined in a namespace and when started the VMI that is passed to the mutating-webhook lacked the namespace.  I'll post another PR with two fixes [1,2] for this and for another memory-related issue I faced.

As for this patch, IIUC the failures in the CI are unrelated.

[1] https://github.com/ahadas/kubevirt/commit/3b86aa32336648a19dd29832dcf8b515654cd8c4
[2] https://github.com/ahadas/kubevirt/commit/9a4fe9ebfa8e37e46d8bbd08076476c8bbd3fbd1

**Comment by ahadas** [comment]
@rmohr can you please follow-up on this?

**Comment by rmohr** [comment]
@ahadas sorry, not right now. The CI issues last week ate a lot of time and other duties do not allow it at the moment. Will come back as soon as possible. Maybe someone else has time.

**Comment by ahadas** [comment]
> @ahadas sorry, not right now. The CI issues last week ate a lot of time and other duties do not allow it at the moment. Will come back as soon as possible. Maybe someone else has time.

Ack.
@cynepco3hahue ?

**Comment by cynepco3hahue** [comment]
travis unhappy with `go fmt`, I think you can run `make bazel-build` and it will format the code accordingly.

**Comment by ahadas** [comment]
@cynepco3hahue done, thanks

File: pkg/virt-api/webhooks/mutating-webhook/namespace-limits.go [hunk 1]
```diff
@@ -21,57 +21,95 @@ package mutating_webhook
...
-
-		namespaceMemLimit, err := getNamespaceLimits(vmi.Namespace, limitrangeInformer)
-		if err == nil && !namespaceMemLimit.IsZero() {
-			if vmiResources.Limits == nil {
-				vmi.Spec.Domain.Resources.Limits = make(k8sv1.ResourceList)
-			}
+	// Copy namespace limits (if exist) to the VM spec
+	if isResourceRequirementMissing(vmi.Spec.Domain.Resources) {
+		defaultRequirements, err := getDefaultRequirements(vmi.Namespace, limitrangeInformer)
+		if err == nil {
```
[the remark this answers is not in the document]
Comment (ahadas) [reply pkg/virt-api/webhooks/mutating-webhook/namespace-limits.go]: done

File: pkg/virt-api/webhooks/mutating-webhook/namespace-limits.go [hunk 2]
```diff
@@ -21,57 +21,95 @@ package mutating_webhook
...
-		for _, val := range limit.(*k8sv1.LimitRange).Spec.Limits {
-			mem := val.Default.Memory()
-			if val.Type == k8sv1.LimitTypeContainer {
-				if !mem.IsZero() {
-					if finalLimit.IsZero() != (mem.Cmp(*finalLimit) < 0) {
-						finalLimit = mem
-					}
-				}
+		// it seems stupid but that's what kubernetes does - it probably assumes a single setting of defaults
+		defaultRequirements = defaultContainerResourceRequirements(limit.(*k8sv1.LimitRange))
```
[the remark this answers is not in the document]
Comment (ahadas) [reply pkg/virt-api/webhooks/mutating-webhook/namespace-limits.go]: the thing is that this code should take the default requirements from, possibly, several LimitRange-s. for instance, one LimitRange may define the default memory limit while default cpu limit is defined in a different LimitRange. changed the code to be more similar to Kubernetes, it is more clear the way they do it

**Comment by cynepco3hahue** [comment]
ci test please

**Pull request merged**